File: RunnersCircleAndroid/app/src/main/java/com/keeganjebb/runnerscircleandroid/support/RunGroupManager.kt
package com.keeganjebb.runnerscircleandroid.support

import com.keeganjebb.runnerscircleandroid.model.RunGroup
import com.keeganjebb.runnerscircleandroid.model.Runner
import com.google.android.gms.maps.model.LatLng
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*


interface RunGroupFragmentInterface {
    fun updatedRunners(aRunners: List<Runner>)
}

interface RunGroupActivityInterface {
    fun notifyDatabaseDisconnect()
}

var runGroupManager: RunGroupManager? = null //Only 1 global instance

class RunGroupManager (aRunGroup: RunGroup) {

    private val mAuth = FirebaseAuth.getInstance()

    private lateinit var firebaseListener: ChildEventListener

    private val mRunGroup = aRunGroup

    private val mRunnersMap = mutableMapOf<String, Runner>()

    private var mRunGroupFragmentListener: RunGroupFragmentInterface? = null

    private var mRunGroupActivityListener: RunGroupActivityInterface? = null

    var databaseReadPeriod: Int = 4 //Number of seconds between database reads

    fun ownerID(): String {
        return mRunGroup.ownerID
    }


    fun userID(): String? {
        return mAuth.currentUser?.uid
    }


    fun runGroupName(): String {
        return mRunGroup.name
    }


    fun setFragmentListenerAs(newListener: RunGroupFragmentInterface?) {

        mRunGroupFragmentListener = newListener
        mRunGroupFragmentListener?.updatedRunners(mRunnersMap.values.toList())
    }


    fun setActivityListenerAs(newListener: RunGroupActivityInterface?) {
        mRunGroupActivityListener = newListener
    }


    fun loadActiveRunners() {

        val groupId = mRunGroup.fireBaseID

        val myRef = FirebaseDatabase.getInstance().reference

        val membersDB = myRef.child(FirebaseString.MEMBERS.value).child(groupId)

        membersDB.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {}

            override fun onDataChange(membersSnapshot: DataSnapshot) {

                for (member: DataSnapshot in membersSnapshot.children) {

                    val memberId = member.key as String

                    val lat = member.child(FirebaseString.LATITUDE.value).value as? Double
                    val long = member.child(FirebaseString.LONGITUDE.value).value as? Double
                    val dist = member.child(FirebaseString.DISTANCE.value).value as? Double
                    val pace = member.child(FirebaseString.AVG_PACE.value).value as? Double

                    var latLng: LatLng? = null

                    if (lat != null && long != null) {
                        latLng = LatLng(lat, long)
                    }

                    val runner = mRunnersMap[memberId]

                    if (runner != null) { //If already active runner, update their object with new data

                        if ((runner.location != null && latLng == null) || !runner.connected) {

                            runner.connected = true

                            val userGroupsDB = myRef.child(FirebaseString.USERS.value).child(memberId).child(FirebaseString.GROUPS.value).child(groupId)

                            userGroupsDB.addListenerForSingleValueEvent(object : ValueEventListener {
                                override fun onCancelled(p0: DatabaseError) {}

                                override fun onDataChange(userGroupSnapshot: DataSnapshot) {

                                    val groupRuns = (userGroupSnapshot.value as Long).toInt()
                                    runner.numRunsInGroup = groupRuns
                                }
                            })
                        }

                        runner.location = latLng

                        if (runner.location != null) {

                            runner.distance = dist
                            runner.averagePace = pace
                        }

                        mRunGroupFragmentListener?.updatedRunners(listOf(runner))

                    } else { //If not already an active runner, add them to the active list

                        val userDB = myRef.child(FirebaseString.USERS.value).child(memberId)

                        userDB.addListenerForSingleValueEvent(object : ValueEventListener {
                            override fun onCancelled(p0: DatabaseError) {}

                            override fun onDataChange(userSnapshot: DataSnapshot) {

                                val userName = userSnapshot.child(FirebaseString.NAME.value).value as String

                                val groupRuns = (userSnapshot.child(FirebaseString.GROUPS.value).child(groupId).value as Long).toInt()

                                val newRunner = Runner()
                                newRunner.runnerID = memberId
                                newRunner.name = userName
                                newRunner.numRunsInGroup = groupRuns
                                newRunner.location = latLng
                                newRunner.distance = dist
                                newRunner.averagePace = pace

                                mRunnersMap[memberId] = newRunner

                                mRunGroupFragmentListener?.updatedRunners(listOf(newRunner))
                            }
                        })
                    }
                }
            }
        })
    }


    fun monitorForRemovedRunners() {

        val groupId = mRunGroup.fireBaseID

        val myRef = FirebaseDatabase.getInstance().reference

        val membersDB = myRef.child(FirebaseString.MEMBERS.value).child(groupId)

        firebaseListener = membersDB.addChildEventListener(object : ChildEventListener {
            override fun onCancelled(p0: DatabaseError) {}
            override fun onChildMoved(p0: DataSnapshot, p1: String?) {}
            override fun onChildChanged(p0: DataSnapshot, p1: String?) {}
            override fun onChildAdded(p0: DataSnapshot, p1: String?) {}

            override fun onChildRemoved(memberRemovedSnapshot: DataSnapshot) {

                val memberId = memberRemovedSnapshot.key as String

                if (memberId == userID()) {

                    mRunGroupActivityListener?.notifyDatabaseDisconnect()

                } else {

                    mRunnersMap[memberId]?.let {removedRunner ->

                        removedRunner.location = null
                        removedRunner.connected = false

                        mRunGroupFragmentListener?.updatedRunners(listOf(removedRunner))
                    }
                }
            }
        })
    }



    fun uploadUserRunData(latLng: LatLng, distance: Double, pace: Double?) {

        val uid = userID()

        if (uid != null) {

            val groupId = mRunGroup.fireBaseID

            val myRef = FirebaseDatabase.getInstance().reference

            val membersDB = myRef.child(FirebaseString.MEMBERS.value).child(groupId).child(uid)

            val runDataMap = mapOf(
                    FirebaseString.ACTIVE.value to true,
                    FirebaseString.LATITUDE.value to latLng.latitude,
                    FirebaseString.LONGITUDE.value to latLng.longitude,
                    FirebaseString.DISTANCE.value to distance,
                    FirebaseString.AVG_PACE.value to pace)

            membersDB.onDisconnect().removeValue()
            membersDB.updateChildren(runDataMap)
        }
    }


    fun removeRunningData() {

        val uid = userID()

        if (uid != null) {

            val groupId = mRunGroup.fireBaseID

            val myRef = FirebaseDatabase.getInstance().reference

            val membersDB = myRef.child(FirebaseString.MEMBERS.value).child(groupId).child(uid)

            val runDataMap = mapOf<String, Any?>(
                    FirebaseString.LATITUDE.value to null,
                    FirebaseString.LONGITUDE.value to null,
                    FirebaseString.DISTANCE.value to null,
                    FirebaseString.AVG_PACE.value to null)

            membersDB.updateChildren(runDataMap)
        }
    }


    fun incrementRunsInGroup() {

        val uid = userID()

        if (uid != null) {

            mRunnersMap[uid]?.let {user ->

                user.numRunsInGroup++

                val groupId = mRunGroup.fireBaseID

                val myRef = FirebaseDatabase.getInstance().reference

                val userGroupDB = myRef.child(FirebaseString.USERS.value).child(uid).child(FirebaseString.GROUPS.value).child(groupId)

                userGroupDB.setValue(user.numRunsInGroup)
            }
        }
    }


    fun removeRunnerFromGroup() {

        val uid = userID()

        if (uid != null) {

            val groupId = mRunGroup.fireBaseID

            val myRef = FirebaseDatabase.getInstance().reference

            val membersDB = myRef.child(FirebaseString.MEMBERS.value).child(groupId)

            membersDB.removeEventListener(firebaseListener)

            membersDB.child(uid).removeValue()
        }
    }


    fun addUserBackToGroup() {

        val uid = userID()

        if (uid != null) {

            val groupId = mRunGroup.fireBaseID

            val myRef = FirebaseDatabase.getInstance().reference

            val membersDB = myRef.child(FirebaseString.MEMBERS.value).child(groupId).child(uid).child(FirebaseString.ACTIVE.value)

            membersDB.onDisconnect().removeValue()
            membersDB.setValue(true)

            loadActiveRunners()
        }
    }


}

File: Pack Track/Pack Track/Supporting Modules/RunGroupManager.swift
//
//  RunGroupManager.swift
//  Pack Track
//
//  Created by Keegan Jebb on 2018-06-15.
//  Copyright © 2018 Keegan Jebb. All rights reserved.
//

import Foundation
import Firebase
import CoreLocation

protocol RunGroupManagerDelegate {
    
    func updateRunners(_ runners: [Runner])
    func notifyDatabaseDisconnect()
}

var runGroupManager: RunGroupManager? //Only 1 global instance

class RunGroupManager {
    
    private var runGroup: RunGroup
    
    private var runnersDictionary = Dictionary<String, Runner>()
    
    private var delegateVC: RunGroupManagerDelegate?
    
    var databaseReadPeriod: Int = 4 //Number of seconds between database reads
    
    init(_ group: RunGroup) {
        runGroup = group
    }
    
    
    //MARK: - Get/Set Private Variable Methods
    
    func ownerID() -> String {
        return runGroup.ownerID
    }
    
    
    func userID() -> String? {
        return Auth.auth().currentUser?.uid
    }
    
    
    func runGroupName() -> String {
        return runGroup.name
    }
    
    
    func setDelegateAs(_ delegate: RunGroupManagerDelegate?) {
        
        delegateVC = delegate
        delegateVC?.updateRunners(Array(runnersDictionary.values))
    }
    
    
    //MARK: - Adding and Retrieving from Firebase
    
    func loadActiveRunners() { //Refreshes changes to active members
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let membersDB = ref.child(FireBaseString.members.rawValue).child(groupID)
        
        //Get the userIDs of all the active users
        membersDB.observeSingleEvent(of: .value) { (membersSnapshot) in
            
            for member in membersSnapshot.children.allObjects as! [DataSnapshot] {
                
                let memberData = member.value as! [String : AnyObject]
                
                let memberID = member.key
                
                let lat = memberData[FireBaseString.latitude.rawValue] as? Double
                let long = memberData[FireBaseString.longitude.rawValue] as? Double
                let coord = (lat != nil && long != nil) ? CLLocationCoordinate2D(latitude: lat!, longitude: long!) : nil
                let dist = memberData[FireBaseString.distance.rawValue] as? Double
                let pace = memberData[FireBaseString.avgPace.rawValue] as? Double
                
                if let runner = self.runnersDictionary[memberID] { //If already active runner, update their object with new data
                    
                    if (runner.location != nil && coord == nil) || !runner.connected { //If it is the first read since finishing a run
                        
                        runner.connected = true
                        
                        let userGroupDB = ref.child(FireBaseString.users.rawValue).child(memberID).child(FireBaseString.groups.rawValue).child(groupID)
                        
                        //Get the most recent value of runs in group for the runner
                        userGroupDB.observeSingleEvent(of: .value) { (userGroupSnapshot) in
                            
                            let groupRuns = userGroupSnapshot.value as! Int
                            runner.numRunsInGroup = groupRuns //Note, this might happen after the delegateVC is updated about the runner. Then it would show this after the next database read
                        }
                    }
                    
                    runner.location = coord
                    
                    //Will not overwrite a value with nil unless they have started running
                    if (runner.location != nil) {
                        
                        runner.distance = dist
                        runner.averagePace = pace
                    }
                    
                    self.delegateVC?.updateRunners([runner])
                    
                } else { //If not already an active runner, add them to the active list
                    
                    let userDB = ref.child(FireBaseString.users.rawValue).child(memberID)
                    
                    //Get the name of each active user and add the information to local runners dictionary
                    userDB.observeSingleEvent(of: .value) { (userSnapshot) in
                        
                        let userSnapshotValue = userSnapshot.value as! [String : AnyObject]
                        
                        let userName = userSnapshotValue[FireBaseString.name.rawValue] as! String
                        
                        let groupRuns = userSnapshotValue[FireBaseString.groups.rawValue]![groupID] as! Int
                        
                        let newRunner = Runner() //Initializes connected to true
                        newRunner.runnerID = memberID
                        newRunner.name = userName
                        newRunner.numRunsInGroup = groupRuns
                        newRunner.location = coord
                        newRunner.distance = dist
                        newRunner.averagePace = pace
                        
                        self.runnersDictionary[memberID] = newRunner
                        
                        self.delegateVC?.updateRunners([newRunner])
                    }
                }
            }
        }
    }
    
    
    func monitorForRemovedRunners() {
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let membersDB = ref.child(FireBaseString.members.rawValue).child(groupID)
        
        membersDB.observe(.childRemoved) { (memberRemovedSnapshot) in
            
            let memberID = memberRemovedSnapshot.key
            
            if memberID == self.userID() {
                
                self.delegateVC?.notifyDatabaseDisconnect()
                
            } else {
                
                if let removedRunner = self.runnersDictionary[memberID] {
                    
                    removedRunner.location = nil
                    removedRunner.connected = false
                    
                    self.delegateVC?.updateRunners([removedRunner])
                }
            }
        }
    }
    
    
    func uploadUserRunData(location: CLLocationCoordinate2D, distance: Double, pace: Double?) {
        
        guard let userID = userID() else {return}
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let membersDB = ref.child(FireBaseString.members.rawValue).child(groupID).child(userID)
        
        let uploadPace: Any = pace ?? NSNull()
        
        let runDataDictionary: Dictionary<String, Any> = [FireBaseString.active.rawValue : true,
                                                           FireBaseString.latitude.rawValue : location.latitude,
                                                           FireBaseString.longitude.rawValue : location.longitude,
                                                           FireBaseString.distance.rawValue : distance,
                                                           FireBaseString.avgPace.rawValue : uploadPace]
        
        membersDB.onDisconnectRemoveValue()
        membersDB.updateChildValues(runDataDictionary)
    }
    
    
    func removeRunningData() {
        
        guard let userID = userID() else {return}
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let membersDB = ref.child(FireBaseString.members.rawValue).child(groupID).child(userID)
        
        let removalDictionary: Dictionary<String, Any> = [FireBaseString.latitude.rawValue : NSNull(),
                                                          FireBaseString.longitude.rawValue : NSNull(),
                                                          FireBaseString.distance.rawValue : NSNull(),
                                                          FireBaseString.avgPace.rawValue : NSNull()]
        
        membersDB.updateChildValues(removalDictionary)
    }
    
    
    func incrementRunsInGroup() {
        
        guard let userID = userID() else {return}
        
        guard let user = runnersDictionary[userID] else {return}
        
        user.numRunsInGroup += 1
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let userGroupDB = ref.child(FireBaseString.users.rawValue).child(userID).child(FireBaseString.groups.rawValue).child(groupID)
        
        //Use the reference to the group to add the new number of runs
        userGroupDB.setValue(user.numRunsInGroup)
    }
    
    
    func removeRunnerFromGroup() {
        
        guard let userID = userID() else {return}
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let membersDB = ref.child(FireBaseString.members.rawValue).child(groupID)

        membersDB.removeAllObservers()
        
        membersDB.child(userID).removeValue() //This cleans up the database so reads only include active runners
    }
    
    
    //MARK: - Handle Firebase Timeouts From App Becoming Active
    
    func addUserBackToGroup() { //Called from delegateVC if the user wants to stay in the group after being disconnected
        
        guard let userID = userID() else {return}
        
        let groupID = runGroup.fireBaseID
        
        let ref = Database.database().reference()
        
        let membersDB = ref.child(FireBaseString.members.rawValue).child(groupID).child(userID).child(FireBaseString.active.rawValue)
        
        membersDB.onDisconnectRemoveValue()
        membersDB.setValue(true) //Change the value in the members dictionary to true to indicate they are active in the group
       
        loadActiveRunners()
    }
    
    
}
